codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,355 @@
+// expected and got wireTypes for a specific tag number.
+type errMismatchedWireType struct {
+	Type         string
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+// String implements fmt.Stringer.
+func (mwt *errMismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched %q: {TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.Type, mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+// Error implements the error interface.
+func (mwt *errMismatchedWireType) Error() string {
+	return mwt.String()
+}
+
+var _ error = (*errMismatchedWireType)(nil)
+
+func wireTypeToString(wt protowire.Type) string {
+	switch wt {
+	case 0:
+		return "varint"
+	case 1:
+		return "fixed64"
+	case 2:
+		return "bytes"
+	case 3:
+		return "start_group"
+	case 4:
+		return "end_group"
+	case 5:
+		return "fixed32"
+	default:
+		return fmt.Sprintf("unknown type: %d", wt)
+	}
+}
+
+// errExtraneousField represents an error indicating that we encountered
+// a field that isn't available in the target proto.Message.
+type errExtraneousField struct {
+	Type     string
+	TagNum   protowire.Number
+	WireType protowire.Type
+}
+
+// String implements fmt.Stringer.
+func (twt *errExtraneousField) String() string {
+	return fmt.Sprintf("errExtraneousField %q: {TagNum: %d, WireType:%q}",
+		twt.Type, twt.TagNum, wireTypeToString(twt.WireType))
+}
+
+// Error implements the error interface.
+func (twt *errExtraneousField) Error() string {
+	return twt.String()
+}
+
+var _ error = (*errExtraneousField)(nil)
+
+var (
+	protoFileToDesc   = make(map[string]*descriptor.FileDescriptorProto)
+	protoFileToDescMu sync.RWMutex
+)
+
+// Invoking descriptor.ForMessage(proto.Message.(Descriptor).Descriptor()) is incredibly slow
+// for every single message, thus the need for a hand-rolled custom version that's performant and cacheable.
+func extractFileDesMessageDesc(desc descriptorIface) (*descriptor.FileDescriptorProto, *descriptor.DescriptorProto, error) {
+	gzippedPb, indices := desc.Descriptor()
+
+	protoFileToDescMu.RLock()
+	cached, ok := protoFileToDesc[string(gzippedPb)]
+	protoFileToDescMu.RUnlock()
+
+	if ok {
+		mdesc := cached.MessageType[indices[0]]
+		for _, index := range indices[1:] {
+			mdesc = mdesc.NestedType[index]
+		}
+		return cached, mdesc, nil
+	}
+
+	// Time to gunzip the content of the FileDescriptor and then proto unmarshal them.
+	gzr, err := gzip.NewReader(bytes.NewReader(gzippedPb))
+	if err != nil {
+		return nil, nil, err
+	}
+	protoBlob, err := ioutil.ReadAll(gzr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	fdesc := new(descriptor.FileDescriptorProto)
+	if err := proto.Unmarshal(protoBlob, fdesc); err != nil {
+		return nil, nil, err
+	}
+
+	// Now cache the FileDescriptor.
+	protoFileToDescMu.Lock()
+	protoFileToDesc[string(gzippedPb)] = fdesc
+	protoFileToDescMu.Unlock()
+
+	// Unnest the type if necessary.
```

Comment:
This is duplicated logic from above. Can we avoid that? I think inverting the of where the cache is checked might do that

Reply:
The inversion would create quite long code blocks i.e. have to wait until a cache miss is handled through the gzip extraction. I instead extracted that shared code to a small helper function. Done, thanks.